### DISCUSSION TITLE
c7n automated IAM policy generation for Lambda execution-role

### DISCUSSION THREAD

**Discussion by zackriso** [cloud-custodian/cloud-custodian#7672] [AWS]
Can Custodian, based on the policy created by the user, predict or make an educational guess regarding the IAM permissions needed for the Lambda execution role? More often than not, many organizations tend to slap an admin IAM role to the Lambda execution role, which is not a good practice. Yet knowing exactly which IAM permissions are needed for each c7n-policy can be challenging or tricky. 

if this is possible, then c7n could potentially auto-create an IAM role with the needed policy and attach it to the execution Lambda on the fly.

**Comment by thisisshi** [comment]
currently no: https://github.com/cloud-custodian/cloud-custodian/issues/937

**Reply by ajkerrigan** [reply]
We can't confidently predict that, but the fact that you said "educational guess" makes it a more tractable problem :). The deployment section of the docs provides some [guidelines around minimal IAM setup](https://cloudcustodian.io/docs/deployment.html#iam-setup) that leans more toward "read-only+" access.

I also have a gist [here](https://gist.github.com/ajkerrigan/f7879cdbbb0a3d285567d8e07e26a723) that serves the "educational guess" case based on policies you've written. It comes with a bunch of warnings about not claiming to be 100% accurate, but as a guess and paired with the deployment docs it can be a useful middle ground between playing permissions whack-a-mole and making your custodian role an Admin.
